### PR TITLE
improve ct gc heuristic for nodeport

### PR DESCRIPTION
**Pull request by borkmann**
Much better and cleaner approach than #9366. ~~Still needs some minor cleanups and rebase. WIP, but to show the POC.~~

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9390)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: pkg/signal/signal.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	config = bpf.PerfEventConfig{
+		MapName:      SignalMapName,
+		Type:         bpf.PERF_TYPE_SOFTWARE,
+		Config:       bpf.PERF_COUNT_SW_BPF_OUTPUT,
+		SampleType:   bpf.PERF_SAMPLE_RAW,
+		NumPages:     2,
+		WakeupEvents: 1,
+	}
+
+	channels [SignalTypeMax]chan<- int
```
Comment (joestringer) [inline pkg/signal/signal.go:60]: It'd be nice to type these a bit more strictly, eg with a `type SignalMsg int` and `chan<- SignalMsg`. That would help to document the code as well.

File: pkg/signal/signal.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if channels[sig.Which] != nil {
+		channels[sig.Which] <- int(sig.Data)
+	}
+}
+
+func signalLost(lost *bpf.PerfEventLost, cpu int) {
+}
+
+func signalError(err *bpf.PerfEvent) {
```
Comment (joestringer) [inline pkg/signal/signal.go:83]: Debug log in this case, maybe?

**Review by joestringer — APPROVED** [review]
Minor nits, looks really clean. This mechanism in general looks powerful for some more event-based co-ordination between the kernel and userspace. Nice work!

File: pkg/signal/signal.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "signal")
+
+	config = bpf.PerfEventConfig{
+		MapName:      SignalMapName,
+		Type:         bpf.PERF_TYPE_SOFTWARE,
+		Config:       bpf.PERF_COUNT_SW_BPF_OUTPUT,
+		SampleType:   bpf.PERF_SAMPLE_RAW,
+		NumPages:     2,
```
Comment (joestringer) [inline pkg/signal/signal.go:56]: Is this one page for the perf header and one page for data? Or actually two pages for data?

**Review by brb — APPROVED** [review]
Nice!

It would be good to add a metric to count signals received (can be done in a separate PR).

I wonder how we could test this. Maybe making `SNAT_MAPPING_{MIN,MAX}_PORT` configurable, setting these two accordingly so that the SNAT port range is small, then sending some traffic and expecting it to succeed and the proposed counter value to be > 0.

**Comment by brb** [comment]
Tested manually on AWS with 10 nodes, haven't observed any `DROP_NAT_NO_MAPPING`!

**Comment by borkmann** [comment]
> Nice!
> 
> It would be good to add a metric to count signals received (can be done in a separate PR).

Ok, will add along with Joe's suggestions in a separate PR.

> I wonder how we could test this. Maybe making `SNAT_MAPPING_{MIN,MAX}_PORT` configurable, setting these two accordingly so that the SNAT port range is small, then sending some traffic and expecting it to succeed and the proposed counter value to be > 0.

We could add some ab/wrk testing to nodeport CI to stress test instead of just single curl.

**Comment by borkmann** [comment]
> Tested manually on AWS with 10 nodes, haven't observed any `DROP_NAT_NO_MAPPING`!

Think it's worth backporting these and prior recent changes to 1.6?

File: pkg/signal/signal.go (hunk 2 above)
Comment (borkmann) [reply pkg/signal/signal.go:83]: For the error case, probably a Debug or Warning makes sense given these may not be expected, but for `signalLost()` I'd leave it as is perhaps. We could consider a Debug statement if there is some sort of rate-limiter for the log, but if not then I wouldn't spam the agent log with it.

**Comment by brb** [comment]
> Think it's worth backporting these and prior recent changes to 1.6?

Considering that 1.7 won't be released soon and that some users might be actually trying to use the NodePort BPF and running into the `DROP_NAT_NO_MAPPING` issue, I'd say yes, let's backport it.

File: pkg/signal/signal.go (hunk 3 above)
Comment (borkmann) [reply pkg/signal/signal.go:56]: Yeah turns out we indeed only need 1 here for the data page as our perf code internally adds the + 1 for the control page. Fixed in #9398 as well.

**Pull request merged**